Error including pb.h headers in a MSVC CLR project

Version: v3.6.0.rc1
Language: C++ integrated into a managed C++ CLR project
Using Visual Studio 2017

I have a C++ library using protobufs.
I have enums defined in proto files and use them in my C++ interface
(Say Enums.proto)

I am wrapping my C++ library in a CLR project for access by C#
The CLR based cpp file will need include the Enums.pb.h to use C++ interfaces with enums

The generated Enums.ph.h includes the stubs/common.h

which has this include block:
// TODO(liujisi): Remove the following includes after the include clean-up.
#include <google/protobuf/stubs/logging.h>
#include <google/protobuf/stubs/mutex.h>
#include <google/protobuf/stubs/callback.h>

And stubs/mutex.h includes std [mutex](https://github.com/protocolbuffers/protobuf/blob/v3.6.0rc1/src/google/protobuf/stubs/mutex.h#L33) (as of 3.6.0.rc1) which is not allowed to be included when building a CLR project.

Error received:

fatal error C1189: #error:  mutex is not supported when compiling with /clr or /clr:pure.

Maybe the include clean-up will fix this if stubs/mutex is not part of common.h
(also once.h includes mutex too)

Maybe a pimpl pattern could be used in WrappedMutex so it does not include std mutex (?)

Yes stubs/once.h also [includes mutex](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/stubs/once.h#L81)
 to gain access to std::call_once, to get CLR compatibility, we would need to move those inline functions using std::call_once out of the header

There are still other headers that include stubs/mutex.h (such as descriptor.h)
and stubs/mutex.h still includes the std mutex header.
For CLR compatibility, we would need to take the std mutex include out of stubs/mutex.h 

Those 2 files are the only ones I see including the std mutex header

I will see if I can possibly remove that include from those files and see how things act

On a side note...
 
Your branch has a few oddities I am working out.
For some reason one of my enums in one of my pb.h files is throwing this error when one of my regular cpp files includes it.
(All my other enums are working fine, so I suppose some generated header chain is broken?)

`
 error C2589: '(': illegal token on right side of '::'
 error C2062: type 'unknown-type' unexpected
 error C3805: 'type': unexpected token, expected either '}' or a ','
 warning C4003: not enough arguments for function-like macro invocation 'max'
`

due to this code

`
enum eHeartRhythm {
  NormalSinus = 0,
  Asystole = 1,
  eHeartRhythm_INT_MIN_SENTINEL_DO_NOT_USE_ = std::numeric_limits<::google::protobuf::int32>::min(),
  eHeartRhythm_INT_MAX_SENTINEL_DO_NOT_USE_ = std::numeric_limits<::google::protobuf::int32>::max()
};
`

--- a/src/google/protobuf/compiler/python/python_generator.h
+++ b/src/google/protobuf/compiler/python/python_generator.h
@@ -38,6 +38,7 @@
 #include <string>
 
 #include <google/protobuf/compiler/code_generator.h>
+#include <google/protobuf/stubs/mutex.h>
 
 #include <google/protobuf/port_def.inc>
 

--- a/src/google/protobuf/stubs/bytestream.cc
+++ b/src/google/protobuf/stubs/bytestream.cc
@@ -33,6 +33,8 @@
 #include <string.h>
 #include <algorithm>
 
+#include <google/protobuf/stubs/logging.h>
+
 namespace google {
 namespace protobuf {
 namespace strings {

--- a/src/google/protobuf/stubs/common.cc
+++ b/src/google/protobuf/stubs/common.cc
@@ -31,11 +31,7 @@
 // Author: kenton@google.com (Kenton Varda)
 
 #include <google/protobuf/stubs/common.h>
-#include <google/protobuf/stubs/once.h>
-#include <google/protobuf/stubs/status.h>
-#include <google/protobuf/stubs/stringpiece.h>
-#include <google/protobuf/stubs/strutil.h>
-#include <google/protobuf/stubs/int128.h>
+
 #include <errno.h>
 #include <sstream>
 #include <stdio.h>
@@ -53,6 +49,15 @@
 #if defined(__ANDROID__)
 #include <android/log.h>
 #endif
+
+#include <google/protobuf/stubs/callback.h>
+#include <google/protobuf/stubs/logging.h>
+#include <google/protobuf/stubs/mutex.h>
+#include <google/protobuf/stubs/once.h>
+#include <google/protobuf/stubs/status.h>
+#include <google/protobuf/stubs/stringpiece.h>
+#include <google/protobuf/stubs/strutil.h>
+#include <google/protobuf/stubs/int128.h>
 
 #include <google/protobuf/port_def.inc>
 

--- a/src/google/protobuf/stubs/common.h
+++ b/src/google/protobuf/stubs/common.h
@@ -47,11 +47,6 @@
 #include <google/protobuf/stubs/macros.h>
 #include <google/protobuf/stubs/platform_macros.h>
 
-// TODO(liujisi): Remove the following includes after the include clean-up.
-#include <google/protobuf/stubs/logging.h>
-#include <google/protobuf/stubs/mutex.h>
-#include <google/protobuf/stubs/callback.h>
-
 #ifndef PROTOBUF_USE_EXCEPTIONS
 #if defined(_MSC_VER) && defined(_CPPUNWIND)
   #define PROTOBUF_USE_EXCEPTIONS 1

--- a/src/google/protobuf/stubs/common_unittest.cc
+++ b/src/google/protobuf/stubs/common_unittest.cc
@@ -31,8 +31,10 @@
 // Author: kenton@google.com (Kenton Varda)
 
 #include <vector>
+#include <google/protobuf/stubs/callback.h>
 #include <google/protobuf/stubs/casts.h>
 #include <google/protobuf/stubs/common.h>
+#include <google/protobuf/stubs/logging.h>
 #include <google/protobuf/stubs/strutil.h>
 #include <google/protobuf/stubs/substitute.h>
 

--- a/src/google/protobuf/stubs/int128.cc
+++ b/src/google/protobuf/stubs/int128.cc
@@ -34,6 +34,8 @@
 #include <ostream>  // NOLINT(readability/streams)
 #include <sstream>
 
+#include <google/protobuf/stubs/logging.h>
+
 #include <google/protobuf/port_def.inc>
 
 namespace google {

--- a/src/google/protobuf/stubs/statusor.cc
+++ b/src/google/protobuf/stubs/statusor.cc
@@ -30,6 +30,8 @@
 
 #include <google/protobuf/stubs/statusor.h>
 
+#include <google/protobuf/stubs/logging.h>
+
 namespace google {
 namespace protobuf {
 namespace util {

--- a/src/google/protobuf/stubs/stringpiece.cc
+++ b/src/google/protobuf/stubs/stringpiece.cc
@@ -35,6 +35,8 @@
 #include <string>
 #include <ostream>
 
+#include <google/protobuf/stubs/logging.h>
+
 namespace google {
 namespace protobuf {
 std::ostream& operator<<(std::ostream& o, StringPiece piece) {

--- a/src/google/protobuf/stubs/stringprintf.cc
+++ b/src/google/protobuf/stubs/stringprintf.cc
@@ -36,7 +36,9 @@
 #include <stdarg.h> // For va_list and related operations
 #include <stdio.h> // MSVC requires this for _vsnprintf
 #include <vector>
+
 #include <google/protobuf/stubs/common.h>
+#include <google/protobuf/stubs/logging.h>
 
 namespace google {
 namespace protobuf {

--- a/src/google/protobuf/stubs/strutil.cc
+++ b/src/google/protobuf/stubs/strutil.cc
@@ -40,6 +40,7 @@
 #include <iterator>
 #include <limits>
 
+#include <google/protobuf/stubs/logging.h>
 #include <google/protobuf/stubs/stl_util.h>
 
 #ifdef _WIN32

--- a/src/google/protobuf/stubs/substitute.cc
+++ b/src/google/protobuf/stubs/substitute.cc
@@ -31,6 +31,8 @@
 // Author: kenton@google.com (Kenton Varda)
 
 #include <google/protobuf/stubs/substitute.h>
+
+#include <google/protobuf/stubs/logging.h>
 #include <google/protobuf/stubs/strutil.h>
 #include <google/protobuf/stubs/stl_util.h>
 

--- a/src/google/protobuf/testing/file.cc
+++ b/src/google/protobuf/testing/file.cc
@@ -46,6 +46,7 @@
 #include <errno.h>
 
 #include <google/protobuf/stubs/io_win32.h>
+#include <google/protobuf/stubs/logging.h>
 
 namespace google {
 namespace protobuf {

--- a/src/google/protobuf/testing/googletest.h
+++ b/src/google/protobuf/testing/googletest.h
@@ -37,6 +37,7 @@
 #include <map>
 #include <vector>
 #include <google/protobuf/stubs/common.h>
+#include <google/protobuf/stubs/logging.h>
 #include <gmock/gmock.h>
 // Disable death tests if we use exceptions in CHECK().
 #if !PROTOBUF_USE_EXCEPTIONS && defined(GTEST_HAS_DEATH_TEST) && \

--- a/src/google/protobuf/util/internal/datapiece.h
+++ b/src/google/protobuf/util/internal/datapiece.h
@@ -34,6 +34,7 @@
 #include <string>
 
 #include <google/protobuf/stubs/common.h>
+#include <google/protobuf/stubs/logging.h>
 #include <google/protobuf/stubs/stringpiece.h>
 #include <google/protobuf/stubs/statusor.h>
 

--- a/src/google/protobuf/util/internal/field_mask_utility.h
+++ b/src/google/protobuf/util/internal/field_mask_utility.h
@@ -36,6 +36,7 @@
 #include <functional>
 #include <stack>
 
+#include <google/protobuf/stubs/callback.h>
 #include <google/protobuf/stubs/common.h>
 #include <google/protobuf/stubs/stringpiece.h>
 #include <google/protobuf/stubs/status.h>

